Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,363 @@
+                var tuple = MuTuple.Create(t.Item1, t.Item2, t.Item1(thing));
+                // add each of them to the listeners list
+                // if it is the first one, a new List must be created and the object be added
+                // otherwise add it to the list
+                listeners.AddOrUpdate(uniqueId,
+                    // add new list
+                    i =>
+                    {
+                        var l = new List<MuTuple<Func<T, object>, Func<T, Task>, object>>();
+                        l.Add(tuple);
+                        return l;
+                    },
+                    // update list
+                    (i, l) =>
+                    {
+                        l.Add(tuple);
+                        return l;
+                    }
+                );
+            }
+
+            // execute all listeners async
+            await Task.WhenAll(listeners[uniqueId].Select(t => t.Item2(thing)).AsEnumerable());
+            // save changes
+            await thing.SaveChanges();
+        }
+
+        public async sealed override Task OnUserDisconnected(IUser user)
+        {
+            await OnThingRemoved(UserListeners, user.SessionId);
+        }
+
+        public async override Task OnChannelDeleted(IChannel channel)
+        {
+            await OnThingRemoved(ChannelListeners, channel.ChannelId);
+        }
+
+        private async Task OnThingRemoved<T>(ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>, object>>> listeners, int uniqueId) where T : IThing
+        {
+            IList<MuTuple<Func<T, object>, Func<T, Task>, object>> dontCare;
+            listeners.TryRemove(uniqueId, out dontCare);
+            await Task.FromResult((object)null);
+        }
+
+        public async sealed override Task OnUserModified(IUser user)
+        {
+            await OnThingModified(UserListeners, user, user.SessionId);
+        }
+
+        public async override Task OnChannelModified(IChannel channel)
+        {
+            await OnThingModified(ChannelListeners, channel, channel.ChannelId);
+        }
+
+        private async Task OnThingModified<T>(ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>, object>>> listeners, T thing, int uniqueId) where T : IThing, IReadModifyWriteObject
+        {
+            // every time a user is modified, all listening functions wtih changed objects need to be called
+            await Task
+            // get all listening functions of the user
+                .WhenAll(listeners[uniqueId]
+                    // check if function has recently been added (Item3 == null)
+                    // or if the old saved object differs from the newly calculated (and directly saved) one
+                    .Where(t => t.Item3 == null || !t.Item3.Equals((t.Item3 = t.Item1(thing))))
+                    // get all listener functions
+                    .Select(t => t.Item2(thing))
+                );
```

Comment:
Sorry for the nitpicking, but I strongly prefer this style.

```csharp
await Task.WhenAll(listeners[uniqueId].Where(t =>
    {
        var prevstate = t.Item3;
        t.Item3 = t.Item1(thing);
        if (prevstate == null)
            return true; // no previous state, so it's new
        return !prevstate.Equals(t.Item3); // different state ==> fire listener
    }) // on a funny sidenote, your old version was buggy as it would never set a new state for new ones (where the state was null)
    // get all listener functions                <-- also, you're _running_ them, not getting
    .Select(t => t.Item2(thing)));
```

Note how I removed some of the whitespace.

You can (of course) merge the `if` and the `return`, I just chose the more verbose version in this case to show my point: Suddenly it makes sense and is readable, at the cost of just *slightly* more code.

But it's still buggy: The mapping func might return `null`, so the nullcheck is not an appropriate solution here.

---

Tuples are fine for small and obvious stuff, but when you have to define mutable tuples because you're doing SO MUCH with them, things start falling apart. Especially when dealing with delegates, `ItemN` is just awkward and while a comment helps, it can't replace a proper name. So I'd strongly recommend killing the tuples and using classes instead.